Add customprofile view.

IvanAndrukh opened the pull request:
Edit CustomProfile model.
Make migration.
Edit update_method in authentication model.
Edit authentication urls.

## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/secure/RapidBoard.jspa?rapidView=738)


## Code reviewers

- [ ] @dmytrorevak
- [ ] @IvanAndrukh
- [ ] @JjuliaP
- [ ] @makarman
- [ ] @noobhunters
- [ ] @Silent777
- [ ] @Staryj812

### Second Level Review

- [ ] @ssOleg
- [ ] @lhalam

## Summary of issue

Edit CustomProfile model.
Make migration.
Edit update_method in authentication model.
Edit authentication urls.

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

lhalam commented:
resolve conflict

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+
+    def get(self, request, user_id):
+        """
+        Method that handles POST request.
```
replace POST word on the GET =)

Silent777 replied:
yeP)

IvanAndrukh replied:
ok

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+                return HttpResponse(status=404)
+
+        profile = user.customprofile
+        if profile:
```
if we create profile automatically is this check necessary?

IvanAndrukh replied:
deleted post method

redmyt replied:
you can remove this `if` check because the user would have the profile in all possible cases. After that, you can remove 400 status code from docstring.

IvanAndrukh replied:
I completely deleted the POST method. It does not need us. This if was inside POST method.

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+            return JsonResponse(information, status=200)
+        return HttpResponse(status=404)
+
+    def post(self, request):
```
is that method necessary if we create profile automatically?

IvanAndrukh replied:
aga

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+        if not profile_data_validator(data):
+            return HttpResponse(status=400)
+
+        requred_keys = ["user"]
```
maybe we could add this checks to `profile_data_validator`?

IvanAndrukh replied:
deleted post

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+        data = {'user': user,
+                'hobby': data.get('hobby'),
+                'photo': data.get('photo'),
+                'birthday': datetime.date.fromtimestamp(data.get('birthday'))}
```
what will be if the `data.get('birthday')` returns `None`?

IvanAndrukh replied:
chanched

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+                'birthday': datetime.date.fromtimestamp(data.get('birthday'))}
+
+        profile = CustomProfile.create(**data)
+        profile = profile.to_dict()
```
maybe it is a good idea to check has the profile created?

IvanAndrukh replied:
deleted post

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+
+        data = {'hobby': data.get('hobby'),
+                'photo': data.get('photo'),
+                'birthday': datetime.date.fromtimestamp(data.get('birthday'))}
```
what will be if the `data.get('birthday')` returns `None`?

IvanAndrukh replied:
changed

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,121 @@
+        :param request: the accepted HTTP request.
+        :type request: `HttpRequest object`
+
+        :return: response with status code 200 when profile was successfully deactivated
```
add the appropriate docstring

IvanAndrukh replied:
done

redmyt commented on eventually/middlewares/loginrequired.py:
```diff
@@ -42,6 +42,8 @@ def __call__(self, request):
                 return HttpResponse('invalid JSON', status=400)
                 # dont forget about loger
+
+        print(request.user)
```
forget to remove the print =)

IvanAndrukh replied:
done

redmyt commented on eventually/templates/emails/registration.html:
```diff
@@ -2,3 +2,4 @@ <h1>Welcome <i>{{ first_name }}</i>!</h1>
 
 <p>Congratulations, your account almost created</p>
 <p>Please click on the link to <a href="http://{{domain}}/api/v1/user/activate/{{token}}">activate your account</a>.</p>
+<p>You have {{time_left}}left to confirm<p>
```
add space after `{{time_left}}` and add the newline to the end of the file

IvanAndrukh replied:
done

redmyt commented on eventually/tests/unittests/customprofile/testview.py:
```diff
@@ -0,0 +1,112 @@
+"""
```
make clean this file

IvanAndrukh replied:
no

redmyt commented on eventually/utils/validators.py:
```diff
@@ -312,3 +316,23 @@ def comment_data_validator(data):
+    Function that validation incoming request.body
+    :param request_body: data that need to validate.
+    :type data: HttpRequest
+    :return: data if data is valid and `None` if it is not.
```
return True not data

IvanAndrukh replied:
ok

redmyt commented on eventually/utils/validators.py:
```diff
@@ -312,3 +316,23 @@ def comment_data_validator(data):
+    :return: data if data is valid and `None` if it is not.
+    """
+
+    if not string_validator(data.get("hobby")):
```
what happens if one of that field value will be `None`?

IvanAndrukh replied:
chanched

Silent777 commented on eventually/templates/emails/change_password_link.html:
```diff
@@ -1,3 +1,4 @@
-<a href="http://{{domain}}/api/v1/user/forget_password/{{token}}">link</a>.
\ No newline at end of file
+<a href="http://{{domain}}/api/v1/user/forget_password/{{token}}">link</a>.
+<p>You have {{time_left}} time left to confirm</p>
```
new line

IvanAndrukh replied:
ok

Silent777 commented on eventually/tests/unittests/customprofile/testview.py:
```diff
@@ -0,0 +1,112 @@
+    #     self.assertEqual(response.status_code, 201)
+    #     # self.assertEqual(response_dict['hobby'], 'computer games')
+    #     # self.assertEqual(response_dict['photo'], 'image3')
+    #     # self.assertEqual(response_dict['birthday'], None)
```
remove comments

IvanAndrukh replied:
deleted

redmyt commented on eventually/customprofile/migrations/0002_added_relation.py:
```diff
@@ -1,22 +1,28 @@
 
     dependencies = [
-        ('authentication', '0001_initial'),
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
```
seems it may be the error. Check if this line could be here.

IvanAndrukh replied:
done

redmyt commented on eventually/customprofile/views.py:
```diff
@@ -1,10 +1,89 @@
+        :type request: `HttpRequest object`
+
+        :return: response with status code 204 when profile was successfully updated
+                 or response with 400, 403 or 404 failed status code.
```
method returns only 400 or 201 status code =)

IvanAndrukh replied:
done

redmyt commented on eventually/utils/validators.py:
```diff
@@ -312,3 +316,27 @@ def comment_data_validator(data):
+    """
+
+    is_hobby_valid = string_validator(data.get("hobby")) if data.get('hobby') else True
+
```
unnecessary blank line

IvanAndrukh replied:
deleted

IvanAndrukh committed:
Add customprofile view.
 Add time message to confirmation email and update password.
 Edited CustomProfile model.
 Making migrations.
 Edit update_method in authentication model.
 Edit authentication urls.
 Add tests to customprofile view.
